src/architecture/merkledb.md

```diff
@@ -0,0 +1,207 @@
+# Exonum MerkleDB
+
+Exonum MerkleDB is a persistent storage implementation based on
+[RocksDB][rocks-db]. MerkleDB provides APIs to work with merkelized data
+structures.
+
+MerkleDB is an object database. Objects represent the highest abstraction level
+for the [data storage](storage.md). All objects fall into the following groups:
+
+- **blobs** which represent sequences of bytes
+- **root objects** that do not have parents. These objects have UTF-8
+  identifiers, for example, "block", "state". Root objects can contain blob
+  items inside them.
+
+## Root Object Types
+
+Root objects perform the same role as in relational database
+management systems (RDBMSs). However, unlike RDBMS tables,
+all MerkleDB objects internally are implemented as wrappers around key-value
+stores.
+
+Root objects in MerkleDB fall into several types:
+
+- **list**, where either keys or values are stored
+- **map**, where key-value pairs are stored
+- **set** of unique items
+- **entry**, which represents a set containing only one item.
+
+Both keys and values in the wrapped stores are persisted as byte sequences.
+MerkleDB does not natively support operations (matching, grouping, sorting,
+etc.) over separate value fields, as it is the case with other key-value
+storages.
+
+### Key Sorting and Iterators
```

Comment:
I am not sure this subsection works here. First, it is too abstract — it requires three pieces of knowledge to understand the iteration order — what is used as keys (not that clear for lists, and sets), how they are serialized, and what is lexicographical order of that. Isn't that too much to ask from a new user? Second — is it important at this point? Shan't it be documented in the iterators? For example, in KeySetIndex#iter, that the elements are ordered by the lexicographical order of their (SHA-256) hashes.

Reply:
the section moved where more appropriate